implementation/src/main/java/io/smallrye/config/AbstractLocationConfigSourceFactory.java

```diff
@@ -0,0 +1,247 @@
+ * locations separated by a comma and each much represent a valid {@link URI}.
+ * <p>
+ *
+ * The locations comprise a list of URIs which are loaded in order. The following URI schemes are supported:
+ *
+ * <ol>
+ * <li>file or directory</li>
+ * <li>classpath resource</li>
+ * <li>jar resource</li>
+ * <li>http resource</li>
+ * </ol>
+ * <p>
+ *
+ * If a profile is active, the profile resource is only loaded if the unprofiled resource is available in the same
+ * location. This is to keep a consistent loading order and matching with the unprofiled resource. Profiles are not
+ * taken into account if the location is a directory.
+ */
+@Experimental("Load additional config locations")
+public abstract class AbstractLocationConfigSourceFactory implements ConfigSourceFactory {
+    public static final String SMALLRYE_LOCATIONS = "smallrye.config.locations";
+
+    protected abstract String[] getFileExtensions();
+
+    protected abstract ConfigSource loadConfigSource(final URL url, final int ordinal) throws IOException;
+
+    protected ConfigSource loadConfigSource(final URL url) throws IOException {
+        return this.loadConfigSource(url, ConfigSource.DEFAULT_ORDINAL);
+    }
+
+    @Override
+    public Iterable<ConfigSource> getConfigSources(final ConfigSourceContext context) {
+        final ConfigValue value = context.getValue(SMALLRYE_LOCATIONS);
+
+        final List<ConfigSource> configSources = new ArrayList<>();
+        if (value.getValue() != null) {
+            List<URI> resources = newCollectionConverter(new URIConverter(), ArrayList::new).convert(value.getValue());
+            if (resources != null) {
+                for (URI uri : resources) {
+                    if (uri.getScheme() == null || uri.getScheme().equals("file")) {
+                        tryFileSystem(uri, context, configSources);
+                        tryClassPath(uri, context, configSources);
+                    } else if (uri.getScheme().equals("jar")) {
+                        tryJar(uri, context, configSources);
+                    } else if (uri.getScheme().startsWith("http")) {
+                        tryHttpResource(uri, context, configSources);
+                    } else {
+                        throw ConfigMessages.msg.schemeNotSupported(uri.getScheme());
+                    }
+                }
+            }
+        }
+
+        return configSources;
+    }
+
+    private void tryFileSystem(final URI uri, final ConfigSourceContext context, final List<ConfigSource> configSources) {
+        final Path urlPath = Paths.get(uri.getPath());
+        if (Files.isRegularFile(urlPath)) {
+            consumeAsPath(toURL(urlPath.toUri()), new ConfigSourcePathConsumer(this, context, configSources));
+        } else if (Files.isDirectory(urlPath)) {
+            try (DirectoryStream<Path> paths = Files.newDirectoryStream(urlPath, this::validExtension)) {
+                paths.forEach(path -> addConfigSource(path.toUri(), configSources));
```

Comment:
By for-each I was referring to [this](https://docs.oracle.com/javase/8/docs/technotes/guides/language/foreach.html), but this is OK too (basically the same overhead though in one case we have an iterator and in the other we have a lambda).

Reply:
Sorry, for some reason I've got use to just call `.forEach`. I guess I need to rewind that :)